src/test/java/org/jenkinsci/plugins/workflow/steps/WaitForConditionStepTest.java

```diff
@@ -194,6 +195,32 @@
         });
     }
 
+    @Issue("JENKINS-27127")
+    @Test public void semaphore() {
+        story.then(r -> {
+            WorkflowJob p = r.createProject(WorkflowJob.class);
+            p.setDefinition(new CpsFlowDefinition(
+                "def semaphore = 10\n" +
+                "def rand = new Random()\n" +
+                "def branches = [:]\n" +
+                "for (int i = 0; i < 50; i++) {\n" +
+                "  def branch = /b$i/\n" +
+                "  branches[branch] = {\n" +
+                "    waitUntil {if (semaphore > 0) {semaphore--; true} else {false}}\n" +
```

Comment:
Since the CPS VM uses green threads, and the closure uses no continuations, it should run atomically. Therefore there is no need for lower-level Java synchronization primitives.